### ISSUE TITLE
DOC Avoiding to install as root on Linux

### ISSUE THREAD

**Issue by rth** [mesonbuild/meson#9034]
The first line of the [installation docs](https://mesonbuild.com/Getting-meson.html#installing-meson-with-pip) says,

>  Meson is available in the Python Package Index and can be installed with `sudo pip3 install meson` which requires root

Installing packages as root from PyPi is practice that's very much frowned upon, and the solution is often to use virtualenvs rather that using sudo. Sources:
 - The [Python Packaging Guide](https://packaging.python.org/tutorials/installing-packages/#id12) 
   > Due to the way most Linux distributions are handling the Python 3 migration, Linux users using the system Python without creating a virtual environment first should replace the python command in this tutorial with python3 and the python -m pip command with python3 -m pip --user. **Do not run any of the commands in this tutorial with sudo**: if you get a permissions error, come back to the section on creating virtual environments, set one up, and then continue with the tutorial as written.
 - Various most voted Stack Overflow posts: https://stackoverflow.com/a/21056000/1791279, https://stackoverflow.com/questions/15028648/is-it-acceptable-and-safe-to-run-pip-install-under-sudo

It's more or less arbitrary code execution as root. So IMO it might be better not to have this as the recommended installation option for unaware users.

In addition from that sentence it's not clear if it's only `sudo` that requires root (and user is free to install it as non root in a virtualenv) or if meson install additionally does something that requires root, and it needs to be installed as root in any case. This question is answered a few paragraphs below  but still.

I can make a PR to improve the documentation if useful.

**Comment by jpakkane** [comment]
> I can make a PR to improve the documentation if useful.

Please do. It would probably make the most sense to switch the order of those two around. That is, first explain how to install it for the current user and then explain why you'd might need to install it system-wide and then how to do that.

**Comment by eli-schwartz** [comment]
The point of using sudo isn't because "if you don't, it will fail to install". The point of using sudo is because "if you don't, it will install by assuming --user, and then you'll be in real trouble if you try running ninja install".

Unfortunately, you CANNOT run user-installed programs as pkexec root with any reliability, specifically here it will fail because the majority of the program relies on the default user-level environment providing resources to a stub script...

Even more unfortunately, you don't even know when meson is going to be invoked as root, because it does its pkexec thing automatically.

Most unfortunately of all, if you have both a system meson and a user meson, they are probably different versions, so running ninja install and getting pkexec'ed to root will invoke the wrong version of meson and fail with deeply involved internal APIs dumping tracebacks at you.

Building and running the zipapp would be a better idea as it's a single-file executable, but we don't publish those and it's a bit of a complicated way to describe installing meson.

The current docs are specifically designed to pressure people into installing meson as root, not because it's generally a good idea to sudo pip install things, but because it's a terrible idea not to do so in the specific case of meson. Unless you have some idea for making it less dangerous.

**Comment by rth** [comment]
Thanks for the explanations!

> Unfortunately, you CANNOT run user-installed programs as pkexec root with any reliability, specifically here it will fail because the majority of the program relies on the default user-level environment providing resources to a stub script... and only the stub script is invoked by pkexec.
>

If I understand correctly there are multiple use cases,
 1. user installs as non root and build package as non root (for instance in a Python virtualenv) -> should not require root or try to escalate privileges. This is in particular the use case considered for replacing distutils with meson for the scientific Python packages starting with scipy https://github.com/scipy/scipy/issues/13615 For this use-case requiring root is blocker. Given the rise of malicious PyPi packages ([last post 3 days ago](https://jfrog.com/blog/malicious-pypi-packages-stealing-credit-cards-injecting-code/)), when pip installing something from PyPi via mesonpep517 that it would require root (or try to gain root) is highly problematic.  
 2. user tries to install a meson built package into a folder that requires root (for non Python packages?). There I understand that higher privileges or some workaround might be necessary. 
 3. users mixing `pip install --user` with `sudo pip install`, which produces permissions issues or otherwise breaks their Python install. There I would say it's on them, because they should not be using `sudo pip install`  and use a virtualenv instead (as documented in the Python Packaging Guide linked above).

And as far as I understand
 -  use case 1 currently actually works fine, 
 -  but the installation documentation is more written with use case 2 in mind

> Unfortunately, you CANNOT run user-installed programs as pkexec root with any reliability, 
> [..]
> if you have both a system meson and a user meson, they are probably different versions, so running ninja install and getting pkexec'ed to root will invoke the wrong version of meson

In summary, that sounds like the automatic privilege escalation with pkexec (I imagine for users convenience) doesn't work unless one undermines security more by running everything as root? I wonder if the the added user convenience (i.e. not asking to re-run with sudo, and if meson is not installed for root install it) is worth it... 

Please correct if I misunderstand something.

**Comment by eli-schwartz** [comment]
For use case 1, the install docs continue on to say "Only use a user copy of Meson if you do not care about installing projects as root."

In the mesonpep517 case, you definitely don't care about installing projects as root, you just want to use `pip` or `python -m build` to run the wheel hook. But actually in that case, you don't need to install meson at all -- the pep 517 builder will install meson as needed into a temporary virtualenv (which may or may not also include build isolation, i.e. don't use system packages) and run everything in the background for you.

Use case 2 is probably the most common one, at least on Linux or generally Unix (on Windows, "install project" is a complicated topic).

**Comment by rth** [comment]
>  the install docs continue on to say "Only use a user copy of Meson if you do not care about installing projects as root."

Yes, it does indeed, but at the very end. Most users probably never get there (I didn't at first), they just run the command they encounter in the first sentence. I'll open a PR to see if the formulation could be improved a bit to avoid this confusion.

**Comment by rgommers** [comment]
>  But actually in that case, you don't need to install meson at all -- the pep 517 builder will install meson as needed into a temporary virtualenv (which may or may not also include build isolation, i.e. don't use system packages) and run everything in the background for you.

Every contributor to a package using Meson will have to have it installed, local development shouldn't use the `pyproject.toml` (PEP 517) hooks.

> should not require root or try to escalate privileges

The "try to escalate privileges" is also tricky indeed. Running `meson install` does this if you forget to use a prefix when running `meson setup`, even if Meson itself is installed in a virtualenv. On my system (Arch Linux) it gives a popup asking for sudo permissions. Erroring out with a clear message would be a lot safer - something like _"you're trying to install to `/usr/xxx` which requires sudo permissions. If that's what you meant to do, please rerun your command under `sudo`"_.

The very first sentence on the install page could also do with a rephrase:

> Meson is implemented in Python 3, and requires 3.6 or newer. If your operating system provides a package manager, you should install it with that

"install it" can be read as "install Meson" or "install Python 3". I hope it's the former, because that's much better practice than `sudo pip install`. But the sentence after it makes it look like it's "install Python 3".

**Comment by eli-schwartz** [comment]
> "install it" can be read as "install Meson" or "install Python 3". I hope it's the former, because that's much better practice than `sudo pip install`. But the sentence after it makes it look like it's "install Python 3".

It was certainly intended to be the former. I hadn't thought before now of it being interpreted as the latter.
